computation_graph/visualization.py

```diff
@@ -22,39 +26,129 @@ def _get_node_shape(node: base_types.ComputationNode):
                 )
 
-
-def visualize_graph(edges: base_types.GraphType, filename: Text = "topology.png"):
-    pgv_graph = pgv.AGraph(directed=True)
-    _make_graph_from_edges(pgv_graph, edges)
-    pgv_graph.draw(filename, prog="dot")
+    return pgv_graph
+
+
+def _do_add_edge(result_graph):
+    return lambda edge: toolz.pipe(
+        result_graph,
+        curried.do(lambda g: g.add_edge(edge, **edge.attr)),
+    )
+
+
+def _do_add_node(result_graph):
+    return lambda node: toolz.pipe(
+        result_graph,
+        curried.do(lambda g: g.add_node(node, **node.attr)),
+    )
+
+
+def union_graphviz(graphs):
+    return toolz.pipe(
+        graphs,
+        # copy to avoid side effects that influence caller
+        gamla.juxt(gamla.compose_left(toolz.first, pgv.AGraph.copy), curried.drop(1)),
+        gamla.star(
+            gamla.reduce(
+                lambda result_graph, another_graph: toolz.pipe(
+                    another_graph,
+                    # following does side effects on result_graph, that's why we return just(result_graph)
+                    # we assume no parallelization in gamla.juxt
+                    curried.do(
+                        gamla.juxt(
+                            gamla.compose_left(
+                                pgv.AGraph.nodes,
+                                gamla.map(_do_add_node(result_graph)),
+                                tuple,
+                            ),
+                            gamla.compose_left(
+                                pgv.AGraph.edges,
+                                gamla.map(_do_add_edge(result_graph)),
+                                tuple,
+                            ),
+                        ),
+                    ),
+                    gamla.just(result_graph),
+                ),
+            ),
+        ),
+    )
+
+
+def save_graphviz_as_png(filename):
+    return curried.do(lambda pgv_graph: pgv_graph.draw(filename, prog="dot"))
+
+
+def save_graphviz_as_dot(filename):
+    return curried.do(lambda pgv_graph: pgv_graph.write(filename))
+
+
+@toolz.curry
+def computation_trace_to_graphviz(
+    computation_trace: Tuple[
+        Tuple[base_types.ComputationNode, base_types.ComputationResult]
+    ],
+):
+    pgv_graph = pgv.AGraph()
+    for node_with_result in computation_trace:
+        _add_computation_node(pgv_graph, node_with_result[0])
+        pgv_graph.get_node(hash(node_with_result[0])).attr["color"] = "red"
+        pgv_graph.get_node(hash(node_with_result[0])).attr["result"] = str(
+            node_with_result[1].result,
+        )[:200]
+        pgv_graph.get_node(hash(node_with_result[0])).attr["state"] = str(
+            node_with_result[1].state,
+        )[:200]
+
+    return pgv_graph
+
+
+visualize_graph = gamla.compose_left(
+    computation_graph_to_graphviz,
+    save_graphviz_as_png("topology.png"),
+)
+serialize_computation_trace = toolz.compose_left(
+    save_graphviz_as_dot,
+    gamla.before(
```

Comment:
why complicate the composition with this and not change the order ?

gamla.compose_left(gamla.compose_left(), save_graphviz_as_dot)

Reply:
because it won't work.
save_graphviz_as_dot first need to get the file name, only then can it be composed after the graph preparations.

Reply:
this is a pretty common pattern when doing high order composition

Reply:
missed the fact save_graphviz_as_dot needs the file name first